### ISSUE TITLE
mapSync on Workers - and possibly on the main thread

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#2217]
In principle, AFAIU, synchronous blocking APIs are only problematic on the main thread. Web workers can block on operations without really causing any problems - especially since our operations should never block for very long (unlike network operations for example, although synchronous XHR is not a good example because it's old).

We could mirror some APIs into synchronous versions with `[Exposed=DedicatedWorker]`.

Definitely useful - I'd start with just this one:
- mapAsync

<details><summary>EDIT: There are other possible entry points but let's ignore them for now</summary>

Maybe useful but possibly not worth the complexity:
- compilationInfo
- onSubmittedWorkDone
- popErrorScope

Most likely not needed:
- requestAdapter
- requestDevice
- device.lost

</details>

Synchronous map (mapSync) is known to be particularly handy. For example, TensorFlow.js can implement its `dataSync()` method to synchronously read data back from a gpu-backed tensor (even if only available on workers).

Finally, there's no avoiding the fact that poorer forms of synchronous readback are always going to be available (e.g. synchronous canvas readback) so we should at least consider supporting mapSync on workers where it's OK in principle.

----

EDIT:

> A possible direction on this issue is that we just give up and start allowing blocking readbacks in general, even on the main thread. I know we're operating under strong architectural guidance that we don't do this. But in practice it's just forcing people to do terrible workarounds with canvases (reshaping/re-encoding data into 2D rgba8unorm data, writing it into a canvas, and then using existing synchronous readback APIs like toDataURL to read them). Maybe we should get out of the business of trying to force people do to the "right thing" when the wrong thing is already possible, and just provide the primitives. The performance consequences of synchronous readback are not that bad (compared with e.g. synchronous XHR).

_Originally posted by @kainino0x in https://github.com/gpuweb/gpuweb/issues/2217#issuecomment-994227968_

**Comment by kainino0x** [comment]
Some customers are very interested in this, so I'd like to put it on the meeting agenda.

**Comment by shanumantesc** [comment]
Without synchronous access to data, we are running into problems with TFJS on WebGPU backend (See https://github.com/tensorflow/tfjs/issues/1595 which suggests that dataSync might never be supported for WebGPU). 

Our rendering strategy calls TFJS as a part of the render loop (since we might want to run an augmented video feed through a neural network) and requires access to network results synchronously for further processing. Yielding the render loop to wait for TFJS results would require some significant refactoring and also introduces an overhead to preserve the renderer state while waiting for the results. Otherwise, we need to build some hacks like trying to reuse latest available results and hope that they are applicable to the frame being processed.

When rendering is happening on a worker thread, there shouldn't be an issue in synchronously waiting for GPU to finish enqueued tasks (and thus support `dataSync` in TFJS) since the main thread can continue processing its event loop without stalls.

**Comment by kainino0x** [comment]
Related to #1629 since we need to make sure (e.g.) map failures are exposed in a consistent and non-destructive way in both versions (sync and async).

meeting: action item @litherum to discuss internally

**Comment by benvanik** [comment]
Wanted to raise our hand as someone who needs blocking operations on workers as well! ✋ (we're compute-focused and don't have concepts of frames or clearly delineated times when we could yield to the browser, and run all our code in workers)

This class of issues is currently our biggest blocker with using WebGPU in the browser - we've only been able to get things working by using `wgpuDevicePoll(..., /*force_wait=*/true)` available in wgpu-native to both simulate this behavior - but even that's not great.

We do also need onSubmittedWorkDone so long as it's not possible to receive a callback for a submission without first yielding to the browser - we're running our entire program in a worker thread without yielding at any point (as would practically anyone coming from webassembly). In our ideal situation this does not imply blocking but that the callback could be issued from another thread such that we can signal a semaphore that a blocking thread may be waiting on. We really would prefer there to be native semaphores/fences in WebGPU but barring that we need to emulate them and the minimum to do so would be a worker thread we kept around to wait for completion events we used to signal potential waiter threads. Something like:

```c
std::thread event_processing_thread([&]() {
  while (true) {
    // we don't want this to be a spin-loop - we're really just telling the API we'd
    // like callbacks to be made from this thread when there are any pending
    wgpuInstanceProcessEvents(/*blocking=*/true);
  }
});
FenceFutex fence;
wgpuQueueOnSubmittedWorkDone(..., [](WGPUQueueWorkDoneStatus status, void *userdata) {
  fence.signal();
});

wgpuQueueSubmit(....);
// can do other work here, non-blocking

// maybe wait - if already fired then this won't block and otherwise it'll wait for the event
// to be processed on the thread and signal the fence
fence.wait();
```

**Comment by benvanik** [comment]
An interesting discussion in the Dawn matrix room with @austinEng where he had a really nice idea and I wanted to document here:

There may be a single primitive that solves this for javascript/native and workers: allow async operations to take a pointer (element in a SharedArrayBuffer) and a value to have the WebGPU implementation write to that pointer when the callback completes. Browser environments could then use `Atomics.wait` (directly in javascript or via futex emulation in webassembly from emscripten) and native environments could use their system futex/WaitOnAddress/etc. Since browsers all have to implement `Atomics.wait` anyway (if they support workers) there shouldn't be much additional infra required here, and users can easily interact with this mechanism themselves via `Atomics.notify`.

Something like:
```c
WGPU_EXPORT void wgpuBufferMapAsync2(WGPUBuffer buffer, WGPUMapModeFlags mode, size_t offset, size_t size, void* 
 addr, uint32_t value, WGPUBufferMapAsyncStatus* status);
```

A synchronous version can then be written easily by users:
```c
bool wgpuBufferMapSync(WGPUBuffer buffer, WGPUMapModeFlags mode, size_t offset, size_t size) {
  WGPUBufferMapAsyncStatus status;
  uint32_t futex = 0;
  wgpuBufferMapAsync2(buffer, mode, offset, size, &futex, 1, &status);
  // could do other work here, switch to another thread/worker, etc - here just blocking to emulate sync behavior
  futexWait(&futex, 1);
  return checkStatusSuccess(status);  // could be device loss, etc
}
```

If all the async methods (`wgpuDeviceCreateComputePipelineAsync` and `wgpuQueueSubmit`) did this then synchronous versions could be written, or more importantly semi-synchronous versions: an application or framework is free to use multiple threads/workers to coordinate work, watch for signals to perform fencing (always be running one frame ahead, etc) without needing spin-loops. In the browser the main thread could submit work and workers could wait on it without dealing with the transfer of promises or other higher-level constructs while still allowing the main thread to poll for completion.

The other advantage of this approach is that it solves the multiple codebase issue: if two frameworks are used in the same application - even if one is written in javascript and the other compiled via webassembly - there's no need for complex scheduling coordination or cross-language interop goo or global WGPUInstance/WGPUDevice waits as everything is scoped.

If we had this I think all our concerns around synchronization would be solvable in user code (we'd be able to perform async compilation, async and overlapped submission, and sync mapping where required) and there'd be little needed additional in the WebGPU API.

This could be emulated with a spin-loop as in my previous example (callbacks take userdata of the addr/value and perform the write themselves), but not having to have user threads/workers spun up to do this - especially if there was one such thread/worker per framework in an application - would be a real win for both ease of use and resource consumption.

**Comment by benvanik** [comment]
@kainino0x helped clarify that this could be seen as the way to do async when you don't have a top-level event loop - the callbacks would work for main thread javascript or a worker that was yielding, and this would cover everything else

**Comment by kainino0x** [comment]
An unfortunate thing about using atomics is that they require SharedArrayBuffer which requires COOP+COEP: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#security_requirements

If SAB were always available (and COOP+COEP just enabled sharing it between threads), or if atomics were allowed on non-shared arraybuffers, then this would work better.

**Comment by benvanik** [comment]
Darn, good point :/

This signal style would be most useful in native or worker contexts with SharedArrayBuffer available (as you're not compiling pthreads code to wasm unless you have them). I think in the case of workers with private memory a blocking tick method that issued the callbacks on the worker that issues the tick would cover the same use cases, though I think the signal style even if not waitable could be useful for ergonomics. I'm imagining the case of dispatching a lot of work and then mapping several of the results: that becomes nice straight-line code of submit->map->map->map->loop on tick until all/any ready->use mapped resources. With the pure callback approach that'd need the user to do more juggling (and it scales with the number of mapped resources, and with each part of the code mapping the resources, etc). Not as efficient as a futex for low-latency wakes but at the point where you have a single thread in isolation you probably don't care much about that. I think even the SharedArrayBuffer method would need some kind of flush method unless the implementations flush automatically, so this may all require the same primitives regardless of approach - it's just if you don't have Atomics.wait you can't block on the signals and deadlock yourself.

**Comment by kainino0x** [comment]
A possible direction on this issue is that we just give up and start allowing blocking readbacks in general, even on the main thread. I know we're operating under strong architectural guidance that we don't do this. But in practice it's just forcing people to do terrible workarounds with canvases (reshaping/re-encoding data into 2D rgba8unorm data, writing it into a canvas, and then using existing synchronous readback APIs to read them). Maybe we should get out of the business of trying to force people do to the "right thing" when the wrong thing is already possible, and just provide the primitives. The performance consequences of synchronous readback are not that bad (compared with e.g. synchronous XHR).

**Comment by benvanik** [comment]
As someone who participated in the WebGL 1 discussions about the same kind of functionality when trying to do compute/media work there I'd agree with most of that - the abominations we had to create still make me sad, but no lack of functionality eliminated our need to create them. Your example is actually one I had to do as "render this canvas to an animated gif" was a product requirement and I couldn't just say "well reading pixels back is really hard in WebGL" - I just had to build a convincing progress spinner because it took so long :P

What we learned then is that there's definite implementation complexity that comes from this kind of stuff but it's really a tradeoff of _what's difficult to do_ in an implementation vs. _what's impossible to do_ on top of the implementation. Needing users to do some extra work to get the precise behavior they want - knowing each user will want a slightly different behavior - works best when the user is actually able to do that work. IME as long as there is a primitive that allows for asynchronously observable multi-threaded semaphores everything else can be built on top of it - blocking sync, callback-style async, pure polling, or mixed futex-based sync/async. If there's only async callbacks with event-loop driven flushes _or_ synchronous blocking operations then the others can't be (practically) implemented by users and there will always be tension (or whining from people like me :). Requiring the performance-destroying use of [asyncify](https://emscripten.org/docs/porting/asyncify.html) when targeting WebGPU or spinning up multiple workers whose only job is to sit and block on waits both fall into that abomination category (this also relates to discussions around efficient timeline-ordered readback via a hypothetical GPUQueue.readBuffer) - people will still need to do those kind of things regardless of whether the implementations make it easy or efficient, and it'd be in the end-user's (web users, etc) best interest for them to be able to be efficient (fewer workers, less large wired allocations, etc).

**Comment by kainino0x** [comment]
This issue is focused on mapSync so I've retitled the issue, and pasted my comment in the summary about allowing it on main thread too.

**Comment by mrshannon** [comment]
I would also like to raise my hand ✋ in support of this. Due to a large existing codebase (that I cannot change) we have to do the following (all in a single synchronous method inside WebAssembly) to implement buffer readback:

1. Get/create a staging buffer
2. Get/create a command encoder
3. Issue a buffer to buffer copy: from the buffer to read to the staging buffer
4. Submit the command encoder to the default queue
5. Wait for the copy to complete (synchronous wait on `onSubmittedWorkDone` which is not currently possible)
6. Map the staging buffer
7. Wait for mapping to complete (synchronous wait, which is also not possible at the moment)
8. Get and return the mapped range

I realize this is pretty inefficient and I would definitely prefer to submit a group of these at the same time and wait on the first to complete, but I can't break the existing codebase.

It also occurs to me that due to internal synchronization, step 5 should not be needed because the map should internally wait until the copy operation is complete. Because of internal synchronization I am not sure there are too many instances where one would need to explicitly wait on `onSubmittedWorkDone`.

**Comment by kainino0x** [comment]
@mrshannon if mapSync were limited to workers, would you be able to get all of this to run on a background thread (assuming support for transferControlToOffscreen)?

> It also occurs to me that due to internal synchronization, step 5 should not be needed because the map should internally wait until the copy operation is complete. Because of internal synchronization I am not sure there are too many instances where one would need to explicitly wait on `onSubmittedWorkDone`, but I can think of one where it would be useful is to query it's status in order to determine if the buffer mapping is likely to stall and instead do other work.

That's correct. Good point it might be more useful to query onSubmittedWorkDone than to synchronously wait on it. (We would define it so it can never change during a task so it's impossible to spin-wait on it.)

**Comment by mrshannon** [comment]
@kainino0x Yes, my preference is to use a background worker, though until offscreen canvas is universally supported I may have to proxy calls to the main thread.

> We would define it so it can never change during a task so it's impossible to spin-wait on it.

That would be a problem in a worker as a program (especially WebAssembly) may not yield to the event loop very often.

**Comment by kainino0x** [comment]
I think that restriction is important to discourage spin looping behavior, so if it's needed then synchronous wait-for-work-done would be preferred.

**Comment by mrshannon** [comment]
I just ran into a case where it cannot be known ahead of time whether to call `mapAsync` or `mapSync`. The situation is:

1. `mapAsync` is called, in hopes that by the time the mapping is required that it will resolve
2. Sometime later, the mapping is needed *now*, and thus a synchronous wait is required and we would like to call `mapSync` but the buffer state is now `mapping pending` and thus it would be illegal to map again.

The obvious solution is to make it so `mapSync` can be called on a buffer which already had a `mapAsync` called on it as long as the range to be mapped is contained within the range requested in the `mapAsync` call.

Without this we must do a busy wait loop with `process.nextTick()`, waiting for the original `mapAsync` to resolve and set a flag.

**Comment by austinEng** [comment]
Maybe we can have `mapAsync` return

```webidl
interface GPUEvent {
  Promise<void> promise;

  [Exposed=(Worker)] // idk if this is allowed here
  void wait();
}
```

**Comment by kainino0x** [comment]
> I just ran into a case where it cannot be known ahead of time whether to call `mapAsync` or `mapSync`.

This is a good point to consider. Similar to #2085: in general you may want to *start* the work as early as possible, but then *block* on it later.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-2022-02-23#mapsync-on-workers---and-possibly-on-the-main-thread-2217">WGSL meeting minutes 2022-02-23</a></summary>



* KN: not sure how much I want to push on this for 1.0, but some things to consider.
* KN: lot of people who'd like to be able to do sync readbacks. Don't know if we want to design it in. Lot of people porting from WebGL - very difficult to change their code or public API to deal with problems like this.
* BD: we're one of those people at Unity. Because Unity API's GraphicsDevice was old, and synchronous. Async API available, but large codebase of existing projects designed around synchronous readbacks of textures & buffers.
* KR: using WebGL 2.0 synchronous APIs? getBufferSubData?
* BD: not really in hot loops. But there are texture readbacks. People do get lazy, and do e.g. reading storage buffers into the CPU to do fake instancing, etc. Bunch of projects around compute shaders that do that. (Not in WebGL 2). Other backends like Vulkan, Metal to this.
* KR: A major difference between Web and native is that WebGPU is going to be sandboxed in another process. No way to make it fast - milliseconds, not microseconds. This is why the API was designed to be asynchronous. You can use `await` in JavaScript, realize it’s undesirable in C++.... Meet team ran into this: readPixels on every frame. Trying to get them to use the “async” readback that Kelsey and Kai developed in WebGL.
* KR: We can put it in, but you’re going to see minimum 1ms roundtrips for these things. Maybe better to just not put these in now, and let application developers find workarounds for now.
* BD: Can’t implement these functions at all as designed. Can do that in the framework in some places, but not functions [exposed to application], testing, …
* KR: For testing, Emscripten Asyncify. Won’t perform well, bloats code size, but can work for tests.
* MS: similar issue with public facing API that requires sync readback. No way to remove without breaking compat with non-web builds. We know it's slow, but lot better than asyncifying everything. We just removed asyncify from our prod builds because of perf impact.
* JB: background q - WebGL 2 also isolates graphics in its own process. What is WebGL doing that WebGPU isn't?
* KG: it's slow.
* MS: we lose perf, too, in sync buffer readbacks. Sometimes can't work around it.
* KR: what if we implemented it behind a flag? Could folks help us characterize the performance?
* BD: we could test it.
* KG: think it's something we'll have to ask people to do the other way for V1. Think this is the time we're going to try to do this right and ask people to do the porting. Experimentation is valuable though, for example if Dawn wanted to try it. Would be nice to find a different way to satisfy this.
* KN: one thing to consider: sometimes may be able to start your map early and not wait on it until later. Could do better than sync - closer to WebGL where you can give it a hint to do the readback early, but make it explicit. Maybe changes to the API if we wanted to do this in the future? mapAsync and say "now I want to map sync"? Maybe don't need to make changes for it, can be side-by-side. Similar things for pipeline creation - start it, wait, then say we need it synchronously later. API design pattern for both of these?
* KG: I'd be more encouraged about doing this if I hadn't just run into a product that was doing it wrong and unwilling to make changes on their side. Experimentation welcome, but gun-shy about this.
* KN: you can always do this a different way - write into a canvas, then read back from the canvas. Can't stop synchronous canvas readbacks. Could say - you won't get your results from the canvas readback, but think that'd be too disruptive. The spec'd say - readback from canvas doesn't happen until some point later. Don't want to do that, would probably break things people do with canvases. That's the biggest reason I'm hesitant to say that we won't do this - people will probably do workarounds like that.
* KG: I'd rather figure out how to stop those working than add this at this point.
* KN: understood. Lot of things to prevent. WebGPU canvas readback; 2D / WebGL canvas interop; a lot to break.
* KR: most of these interop cases should be async and must not break.
* KN: right. So only the readback case is changing in some way - changing behavior of other specs.
* KG: when you use this canvas as a source, this is what happens. Ill specified? There have been questions about whether ops pull from the front or back buffer in WebGL. If front buffer - it's what rAF picked up last time.
* KN: not just impl details. We can do that for WebGPU, but for any other canvas that uses the WebGPU texture, and any other canvas that uses those. Strange propagation through lots of other canvases.
* KG: no - saying drawImage picks up the last frame.
* KN: don't think that'd work. People rely on things like that working.
* MS: if you do this in JS I don't see a reason to have a sync API. When in Wasm, needs to be a standard way of doing this that isn't asyncify the whole codebase.
* KG: isn't answer Emscripten bindings? WebGL Emscripten bindings weren't written in Wasm. Has a cost, but it's fine.
* KN: can't pop out to async JS.
* MS: that's the issue. 2 places where it hurts: mapAsync, and submitWorkDone. Both of these are sore points in writing WebGPU code in Wasm. Can't wait on it - have to come up with a sync wait on it. For me, it's a bunch of 0-tick callbacks and proxying everything to another thread. Not great way to do it. If a way to address these two pain points in Wasm would be great.
* KG: understood. In some way, there's no way around this. If you have a tight loop of Wasm code, because of our rules, if we don't yield to the event loop we can't give you results back. Said the timeline doesn't work that way. Then no API we can give you unless we change how that timeline works. Can things be considered "done" before you return to the event loop?
* MS: in general - in Wasm you don't return to the event loop until next frame. So, can't map buffers, and can't wait on submitted work, until the next frame. Breaks a lot of stuff.
* KG: we'll keep thinking about it.
* KN: if we have a function that creates 1 ms janks, and people use and complain about it - it's still better, 90% of people needed it and will take the perf tradeoff if they wanted.
* KG: not comfortable with doing this given that I wasn't able to convince a team at a member company here to follow the best practices for the existing API (WebGL 2.0). Soured me on this notion, unfortunate.
* KN: is it slower in your browser than others?
* KG: it is.
* KN: WebGL readbacks?
* KG: yes. Some things we can do to make it better, but I have other things I have to work on. Still disabled in Firefox.
* KN: maybe put it off until we can try to get matching performance.
* KG: considering reverse origin trial. If some origin starts doing poorly performing things we can push them to fix their code. Right now people are choosing a different browser for these things.
* KN: we do interventions for things like this - slightly different.
* KN: let's just have Wasm solve this async problem.
* KG / MS: would be nice.
* KN: some progress has been made, but not sure if any in the past 2 years.


</details>

**Comment by toji** [comment]
The editors were talking about this issue recently, and I got the itch to build software atrocities over the weekend. As a result, I'm proud(?) to present, `mapSync` in WebGPU:

[Implementation](https://github.com/toji/webgpu-map-sync/blob/main/map-sync.js)

Example Usage [(Full sample)](https://github.com/toji/webgpu-map-sync/blob/main/index.html#L92)
```js
import './map-sync.js';

const buffer = device.createBuffer({
  size: 1024,
  usage: GPUBufferUsage.MAP_READ_SYNC | GPUBufferUsage.COPY_DST
});

// Copy data into the buffer

buffer.mapSync(GPUMapMode.READ); // Note the lack of await!
const arrayBuffer = buffer.getMappedRange(8, 256);
console.log('buffer contents: ', new Uint32Array(arrayBuffer));
buffer.unmap();
```

And I sincerely hope nobody ever uses it.

The TL;DR is that this codes take's Kai's comment from earlier entirely too seriously:

> in practice it's just forcing people to do terrible workarounds with canvases (reshaping/re-encoding data into 2D rgba8unorm data, writing it into a canvas, and then using existing synchronous readback APIs like toDataURL to read them)

In order to present an API that looks and works like a proper synchronous mapping it polyfills WebGPU to intercept the `MAP_READ_SYNC` usage, replaces it with `STORAGE` behind the scenes, and then in `getMappedRange()` it:

 -  Draws the desired range of the buffer to into a WebGPU canvas's current texture, packed into `bgra8unorm`.
 -  Uses the WebGPU canvas as a source for `gl.texImage2d()` in a WebGL2 context.
 -  Sets that texture as the color attachment for a framebuffer.
 -  Calls `gl.readPixels()` on the framebuffer to get the data out synchronously.

This _technically_ offers the behavior of reading data back synchronously in JavaScript, but it's stupidly expensive. In my sample page, I read 256 bytes out of a larger buffer with equivalent sync and async code paths.

 - Polyfilled `mapSync` path: 64 - 102ms. (Varies wildly)
 - `mapAsync` path (including `await`): ~16ms.

So _best case scenario_ this Rube Goldberg machine of a library is taking 4x longer and using 3x the memory (probably more) than if we had just let the async mapping happen. I suspect an actual native `mapSync()` would take even less time than the `mapAsync()` shown here because it wouldn't be processing microtasks mid-operation. Obviously, we want to encourage people to use `mapAsync()` anywhere they possibly can, but as we've seen some use cases like @mrshannon's can't accommodate that.

As a result I'm concerned that some developers will look at readback paths like this, recognize that they're terrible, and use them anyway because it turns out to be the only way to accomplish what they need. I don't want anybody to have to use the code I've written here, but honestly there's nothing in it that your average WebGL/WebGPU dev couldn't figure out in a weekend like I did. If I didn't make it someone else would, and in fact several someone else probably still will. The end result being that this technique _will_ end up in use on live pages unless a better alternative is provided by the API itself.

So, this is all a long way of repeating what Kai already said in the opening comment:

> there's no avoiding the fact that poorer forms of synchronous readback are always going to be available (e.g. synchronous canvas readback) so we should at least consider supporting mapSync

I just have proof now of how bad the alternative is.

**Comment by greggman** [comment]
I've been trying to think of arguments for and against synchronous support. I think I already know the arguments against.

* The GPU is managed by a separate process. Synchronizing between the two is slow. 
* The browser's main thread is synchronous and so blocking JavaScript blocks the main thread.

A few arguments in support of synchronous

* WebGL has synchronous readback, both `readPixels` (for 11yrs) and `getBufferSubData` (for 5years) and the world hasn't ended. Most developers don't want janky apps. They don't seem like they are making janky apps with these feature.

* JavaScript itself has the same issue. I can do `for (value of millionValues) heavyFn(value)` and block the browser for up to 30 seconds (when it times out and asks the user if they want to kill the page).

  That's not to say we should add more things that can block the page but it would be nice to know if there is some core principle that suggests what must be async.

   * Is it the case that long operations must be async? No (see JavaScript example above)
   
   * Is it because the GPU is in a separate process? Maybe?
   
     That the GPU is in a separate process is arguably an implementation detail. Some browsers are not multi-process or don't have a GPU process (yet). The GPU itself might be a separate process but swiftshader is not, just to give an example of what could be a same process implementation. The Canvas 2D API started as all CPU in the same process. It's certainly valid to say that the major browsers are all trying to have GPU processes but is that enough reason to mark WebGPU as "must be 100% async"?

   * Is it because talking to the GPU and getting a result is slow?
   
     That's certainly not guaranteed. [I wrote a WebGL2 example to compute the average of a 16 million value Float32Array](https://jsgist.org/?src=f451e3d46e18e3217d58acec7253c757). Even including uploading the 16 million values to the GPU and a sync result readback I found the GPU solution 2x to 16x faster on my 2019 MBP (AMD) and my 2021 M1 MBP. At 16k values the CPU won. At 64k values the GPU won (including uploading and sync readback). Obviously this will vary by browser/OS/GPU.

The point is, maybe sync map isn't going to end the world and maybe most developers will mostly only use it where it is performant enough?

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgpu-web-2022-03-16#mapsync-on-workers---and-possibly-on-the-main-thread--issue-2217">WebGPU meeting minutes 2022-03-16</a></summary>



* BJ: wrote something. Kai and I discussed mapSync. Point came up multiple times - you can already do this with clever combinations of writing to the canvas + existing mechanisms like toDataURL. I tried it. Wrote a comment. Effectively polyfilled mapSync.
* BJ: Jim Blandy's is the appropriate reaction. :)
* BJ: you give it a MAP_READ_SYNC usage. Replaces it with STORAGE. Calling mapSync function - uses shader to write storage buffer contents to canvas. toDataURL, upload to WebGL2 texture, readPixels it back. Smallest memory usage/copies. At least 3x memory usage, probably 4 copies of the data. But, it works! Looks/acts like mapAsync, but synchronous.
* BJ: at least 4x slower, maybe more.
* BJ: overall it's a bad idea. Hope nobody uses it.
* BJ: but we've had feedback from multiple places, including internal partners - code they're porting simply requires sync readbacks, or would take so much rearchitecting to not use it, not worth their time.
* BJ: distinct disadvantage compared to WebGL - which have readPixels, getBufferSubData. A reason people would stay with WebGL when they'd rather update to WebGPU-based code.
* BJ: think we should consider this, if for no other reason - make sure we represent a strict superset of what WebGL offers.
* MM: q: on D3D12, where staging buffers are required - how do they do sync readback?
* KG: you can do it through staging buffers.
* MM: don't you need to schedule a copy from real buffer to staging buffer? Different timeline?
* BJ: to be clear - we have that requirement too. Can only mapRead a staging buffer. Must have copied buffer earlier. No compute shaders on staging textures. Would have had to do the copy to staging buffer before mapSync. Appears to be doable synchronously. Wait for that operation to finish, then surface the data. No different than readPixels in WebGL.
* MS: to answer a different way - we use persistently mapped buffers and GPU fences to wait.
* MM: so you'd need sync map & sync wait for queue completion?
* KG: the sync map is the wait.
* MS: you don't need both.
* KG: you can make one with the other. Think it goes both ways. Enqueue copy to staging; insert fence; wait for fence to complete.
* MM: I get it now.
* BJ: we already do it. Otherwise my polyfill wouldn't be possible.
* KG: there's a lot of nuance to this. Part of the reason we haven't run into this with getBufferSubData is because WebGL2 adoption's low. Have seen readPixels happening; does cause problems. Difficult to do this thing because you're not supposed to. Dissuades people from doing so. Interesting point - too principled to stand by that as a blocker. You might need to do this sometimes, just please don't abuse it.
* BJ: one thing - we do see readPixels in WebGL, can be a perf problem. But, doing the right thing in WebGL with buffer readbacks - not possible until WebGL2. Difficult to do async buffer readbacks in WebGL2. Annoying, non-intuitive. Here, async path's obvious. No concerns pushing that in every way. Think people will use it. Hopefully, we can present the async path as far more attractive than the sync one. But we still do run into some subset of uses where it doesn't work for them.
* MM: agree with Kelsey, things we don't want authors to do should be hard. Even if it's possible for authors to do the hard thing, if they have to write hard code to do so, that's intended. Don't think having mapSync next to mapAsync will be good - it's convenient, people will call it. Think having to jump through APIs is a good thing.
* BJ: think there's a stratosphere - making it hard, needing crafty incantations - not sure I'd want something sitting next to mapAsync. No qualms making it yell at the console on first use. "This is going to be slow." Open to many options to make it harder to reach for. But, at end of day if you still need it - don't want it to take 4x the time, memory, copies than a native impl.
* CW: to make a note - should separate mapSync in main threads from it in workers. Seems OK to have sync APIs in workers. Easier to decide we want it in workers? Having it on the main thread introduces jank in animations, or worse.
* BJ: in general yes - politically easier to put blocking APIs in workers. Have heard some pushback on that in some forums. With Kai have heard - don't want to split the API if we can avoid it. If there's the will to make it work in a worker and not on main - worth a try. May not solve all use cases. There are plenty of cases where it's difficult to do a particular kind of work on a worker. May need keyboard/mouse, poll from video - things that don't work on workers.
* KR: Wanted to advocate for splitting this to something only available on the worker. If this group can come to agreement - there are other synchronous APIs on workers. There have been many discussions with people, especially TensorFlow.js and their users, where they just can’t achieve what they want on WebGPU. And they get a lot of benefits from WebGPU and would like to use it. If we can agree to put this on in DedicatedWorkerGlobalScope then people can experiment with it and we’ll have more input on supporting this generally.
* MM: agree splitting the convo into 2 halves is valuable. Cost/benefit tradeoff is different on workers than main thread. Hope we aren't the first API to add a synchronous function that takes many frames. Hard to get past TAG. But, on workers there are pros and cons.
* MS: concerns about adding this. If you do add mapSync - what will happen is, in Wasm - people including myself, maybe, will use it to get around the fact that we can't yield. And yet, mapAsync to return within a frame. Will cause it to be used a lot more than you intended to get around another issue with WebGPU in Wasm.
* MM: what's the other issue?
* MS: yes. Essentially - other APIs can map, then query when the map's completed. Reason that's an issue - can't do it in Wasm because you aren't yielding. None of buffer maps resolve until next frame unless you proxy things. If mapSync's available, it's an easy but terrible workaround. We'll use it because mapAsync never resolves in the same frame.
* KN: that's a really good point. Thanks for raising it.
* GT: trying to understand guiding principle. Can block JS by iterating through million element array, etc. Why is this required to be async and others not? JS's new findLastIndex can block if you have an expensive callback.
* MM: think the answer is - because the website's written the code that takes a long time to execute. Here, you're blocking on another process. In JS you're not.
* GT: the other process is an implementation detail. There are single-process browsers; there are SW WebGPU impls. Trying here to find out why this has to be async.
* CW: think difference is - some things are algorithmic things you have to do in JS right now. Others are I/O we don't know how long will take. Maybe you wrot ethe WebGPU code - know how long it willt ake on the GPU. Major difference - when you need to do DOM transformations, need to do it now - GPU operations are inherently async. People want sync mapping, because we're used to it in native, and in WebGL! But people don't want to change their code. It's a friction point. There's nothing blocking the developer from updating their code. When doing JS to do DOM updates - need to do them for this frame, so can't wait. That's the difference. JS has to do the update right now - or in the other case it's developer convenience.
* MS: there is a case where it's not possible at all to refactor - we have that, and the other Brandon mentioned they have the same issue in public APIs we can't change without breaking everyone's code.
* CW: in terms of web platform - it's still your code. The code that runs in the browser. All under the developer's control. Agree, amount of friction's large - but it's still friction.
* KR: TF.js feedback is they just can’t get done what they need to get done. Orders of magnitude slower than what they can get with WebGL.
* KG: Hope we [could resolve that]
* MM: Can you link that?
* 
* 
* 
* 
* 


</details>

**Comment by kainino0x** [comment]
In the next meeting, I think we should focus on trying to come to an agreement on mapSync in workers. (`[Exposed=DedicatedWorker]`)

**Comment by litherum** [comment]
See: https://w3ctag.github.io/design-principles/#promises

Even if we did think that this was a good idea (we don't), it's clear that this issue doesn't need to be in V1. It can be added later if necessary.

**Comment by kainino0x** [comment]
> Even if we did think that `mapSync()` was a good idea (we don't)

Even in workers? There is plenty of precedent for blocking operations in workers, like Atomics.wait and synchronous XHR.

> it's clear that this issue doesn't need to be in V1. It can be added later if necessary.

Technically true, but we have input from a lot of customers who will be blocked from using V1 if we don't at least have this in workers, because it's a regression in functionality compared with WebGL.

**Comment by pyu10055** [comment]
We have learned from several important customers of TensorFlow.js, there is a strong need for synchronous reading API for WebGPU within web worker environment. I will list couple examples here:
1. Async call is very hard to communicating with WASM, the current WASM asyncify mechanism posts a significant performance drain.  
2. For real-time applications that utilize vision models, sync read is needed to achieve high performance on model inferencing.
3. TensorFlow.js needs to copy computation results between backends (WebGPU, WASM, WebGL, etc), to execute kernels that are not available in one particular backends. Having a synchronous data read API will significantly simplify the data transfer logic.

**Comment by lisanlow** [comment]
My product needs this mapSync support to read pixel datat from a texture synchronously, asynchronous call is not feasible for us.

**Comment by toji** [comment]
I was just introduced today to the concept of the `FileSystemSyncAccessHandle`, which seems to have some good parallels to `mapSync()`. Specifically, it's a synchronous alternative to the [`getFile()` method](https://fs.spec.whatwg.org/#api-filesystemfilehandle-getfile) of the File System API that is only available in a worker and aimed at higher performance file access, especially from WASM.

That feels like it matches the primary use cases that we're hearing for `mapSync()` pretty well, and if it's acceptable for the File System API I'd suggest that it's equally acceptable for our needs.

It appears to still be a relatively recent API and so resources are sparse, but there are some relevant links if you look around enough.

Links:
 - https://webkit.org/blog/12257/the-file-system-access-api-with-origin-private-file-system/
 - https://groups.google.com/a/chromium.org/g/blink-dev/c/UzGXhVEFpwU
 - https://github.com/w3ctag/design-reviews/issues/772

Additionally, I'd like to point out that I've recently had a conversation with the WebAudio working group, who expressed an interest in performing some audio transforms on the GPU via compute shaders. Audio is highly sensitive to latency, so having mechanisms for reading back the transformed data with as little latency as possible seems like an important feature for them.

**Comment by 94bryanr** [comment]
I'd like to chime in as another user who would love to see more progress on this.

I'm writing a UI library which utilizes a compute shader for bounding box / intersection calculations (on tens of thousands of objects). I need the shader to run and send its results back to the CPU per-frame to avoid introducing lag during user interactions. Using `asyncMap` it takes ~30ms or ~4 frames for the mapping operation to resolve which results in a low quality experience.

Not having a way to synchronously get data off of the GPU with low latency is a blocker for my use case and (I feel) a number of other interesting potential applications.

---

Note:  I'm building using Rusts wgpu API which exposes [a device poll function](https://docs.rs/wgpu/latest/wgpu/struct.Device.html#method.poll). This allows me to run the code on native - where I can see that the synchronous wait results in exactly the behavior I want. So I know the issue here is not in the hardware - just how I can currently make use of it through this API.

I am still investigating ways to reduce the latency through web workers, but I am discouraged by a first attempt at refactoring as it appears `createBuffer` causes a panic when called from a worker - so it will be a non-trivial refactor for me to move my workload. Even if I do, it is not obvious what kind of performance gains I will see.

**Comment by Dinnerbone** [comment]
We also need this for Ruffle. We emulate Flash and, even though it's not at all recommended by modern standards, we need to be able to read back gpu buffers synchronously whenever the script calls for it.

WebGPU requiring async because it's best modern practice, means we can't port anything that predates that advice to switch to WebGPU.

**Comment by greggman** [comment]
IIUC the intended solution for simulating sync reads for wasm is [JSPI](https://v8.dev/blog/jspi)

**Comment by torokati44** [comment]
+1 for the comment above.

**Comment by Kangz** [comment]
> See also [Blink's decision to raise the synchronous WebAssembly.compile() module size limit from 4KB to 8MB](https://groups.google.com/a/chromium.org/g/blink-dev/c/nJw2zwaiJ2s/m/EYPgC5D3LwAJ?pli=1), even if the main thread is blocked for almost a second.

I happened to be chatting to the author of that intent to ship today, he added some color in that if you look at the thread, there is significant pushback: making WebAssembly modules block for 1 second for compilation is terrible for the Web platform. After some negotiation they settled on the 8MB limit that's blocking for much less time (basically a heavy function call, but at the limit of considered "blocking").

It's clear that there is convenience in being able to map synchronously, but JS has a ton of primitive to handle asynchrony, and synchronous readback in the GPU-picking cases you mentioned would easily stall for a couple frames due to the depth of the browser rendering pipeline. For GPU picking it is still better to do async readbacks. Also as an additional data point, playing Baldur's Gate 3 I noticed an extremely slight latency in item picking, which I assume is due to the async GPU picking, but however there is no stall when clicking on something.

**Comment by torokati44** [comment]
>  For GPU picking it is still better to do async readbacks.

That is just one example. And even that is not always acceptable.

> and synchronous readback in the GPU-picking cases you mentioned would easily stall for a couple frames due to the depth of the browser rendering pipeline. 

It would be still preferable to leave this decision up to the developers using the spec, and not simply taking away one of the choices.

While it may not be ideal, in some cases it's a perfectly valid tradeoff. See for example: https://github.com/gpuweb/gpuweb/issues/2217#issuecomment-1637115876

**Comment by greggman** [comment]
An argument I agree with that I've heard against sync is that there should be is no performance different between sync and async. One way or another the GPU process has to signal to the process waiting for the result that the result is ready. After than, you can imagine in a perfect implementation, the difference between sync an async is just a few instructions.  So, if async is slow in some browser, that's an implementation bug, not a reason to add sync.

An exception would be wasm but as mentioned above, [JSPI](https://v8.dev/blog/jspi) is supposed to be the solution for WASM

**Comment by lvyitian** [comment]
> An argument I think I agree with that I've heard against sync is that there should be no performance difference between sync and async. One way or another the GPU process has to signal to the process waiting for the result that the result is ready. After that, you can imagine in a perfect implementation, the difference between sync and async is just a few instructions. So, if async is slow in some browser, that's an implementation bug, not a reason to add sync.
> 
> An exception would be wasm but as mentioned above, [JSPI](https://v8.dev/blog/jspi) is supposed to be the solution for WASM so ports will be easy.
> 
> Picking was brought up as an example and I worried about that too. it's possible it is a problem but I wrote [this example](https://jsgist.org/?src=57c3c0d3cf1db14f1baad2d9b0094397) and it doesn't seem like a problem. The example makes 1000 black circles in SVG that turn red on hover. It also makes 1000 black circles with WebGPU and uses GPU picking to mark one as red. I thought I might be able to see a noticable difference between an SVG circle being highlighted vs WebGPU circle being highlight but I don't. Maybe my eyes just aren't sensitive enough or maybe you need to make a much heavier scene for it to stick out.

According to https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md#supporting-responsive-applications-with-reentrancy
JSPI won't deal with reentrancy automatically, the current project architecture will have to be reconstructed to be "event-loop style" to work properly.

**Comment by valadaptive** [comment]
> JS has a ton of primitive to handle asynchrony

No, it does not. It has callbacks, syntactic sugar over callbacks (promises), and syntactic sugar *over* syntactic sugar over callbacks (async/await). Promises can be implemented entirely as a library, and async/await is a very simple syntactic transformation that's existed since ES6's generators.

You are one of maybe a few dozen people in the world who actually knows what a microtask is and exactly which operations queue them up. Every other developer sees the `async` keyword in front of a function and starts dreading the prospect of rearchitecting their entire rendering engine to use async/await, including bug-filled and half-baked versions of every synchronization mechanism you've used in the WebGPU spec to prevent race conditions.

If JS included *real* async primitives, and by *primitives* I mean language features that are fundamental enough that they cannot be polyfilled, I suspect that fewer people would react the way they do when some extremely useful function is made async because the Web Platform™ gods have deemed it too janky for the uneducated masses to use without potentially causing a 1-2 frame slowdown (the horror!)

> synchronous readback in the GPU-picking cases you mentioned would easily stall for a couple frames due to the depth of the browser rendering pipeline

"Would" or "could"? WebGL's `readPixels` certainly doesn't seem to have this problem. I can call it dozens of times per frame to read data from an offscreen buffer with no slowdown. Is there something fundamental to WebGPU that causes every pipeline to be inherently much deeper than the equivalent WebGL state configuration?

**Comment by Kangz** [comment]
Debating what to qualify `async/await` and other features of Javascript isn't going to help get to a consensus. It makes it possible to write code that does async GPU readbacks without dealing with a mess of callbacks. Of course the rest of the code often needs to be adapted to handle the asynchrony because it might try to do something in the meantime, and there's a development cost to that.

It is clearly an unpopular opinion among folks that are used to WebGL's `readPixels`, but adding blocking calls on the JS main thread is unbelievably difficult to motivate nowadays and extremely frowned upon by Web standards folks and browser implementers. However as discussed in this issue, the same guideline doesn't exist for workers, so there could be a `mapSync` on workers. There will be a lot of details to figure out still.

> "Would" or "could"? WebGL's readPixels certainly doesn't seem to have this problem. I can call it dozens of times per frame to read data from an offscreen buffer with no slowdown. Is there something fundamental to WebGPU that causes every pipeline to be inherently much deeper than the equivalent WebGL state configuration?

Would. It requires all previous operation on the GPU to be completed, and because most GPUs have a single graphics queue, it forces all the compositing of the browser to be finished etc. If you block every frame, then the amount of queued work stays low all the time, but you still create lots of bubbles in the execution. If you look at the performance profile of your page, you should be able to see that each readPixels still take a large amount of time (~1ms at least because of the IPC roundtrip and submit + immediate wait on GPU work). (unless you use [non-blocking async readback](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices#use_non-blocking_async_data_readback) but that's basically like `mapAsync`).

**Comment by litherum** [comment]
(Apologies if this idea has been made already)

What if the native implementations, but not the web implementations, add support for synchronous readback on the main thread?

Pros:
- The web is where the hard requirement to not block the main thread comes from. (Native apps have a similar desire to not block the main thread, but there aren’t as many institutions built up around that as there are around the web not blocking the main thread.)
- Some (but not all) authors get what they want

Cons:
- Authors targeting the web don’t get what they want
- Authors targeting both native and the web still have to pay the cost of no synchronous readback on the main thread

**Comment by valadaptive** [comment]
> synchronous readback in the GPU-picking cases you mentioned would easily stall for a couple frames

> each readPixels still take a large amount of time (~1ms at least...)

"a couple frames" to "~1ms" is quite a big pivot! Even on a high-end phone with a 120hz display, that's still only an 8th of a frame.

I've benchmarked readPixels, and it indeed takes about 1ms per call. Here are some points of comparison ([that you can try for yourself](https://github.com/gpuweb/gpuweb/files/14378935/1ms-benches.zip)) for what else takes 1ms:
- 10000 `Uint8Array#subarray` calls. It's a bit of a microbenchmark: no other operations but `subarray`.
- Updating the positions of 150 SVG `<circle>`s.
- 10 forced reflows.

Does this mean we should make number-crunching code, SVG DOM operations, or [all reflow-forcing operations](https://gist.github.com/paulirish/5d52fb081b3570c81e3a) asynchronous?

1ms is, IMO, absolutely *not* heavy enough to justify forcing async.

**Comment by torokati44** [comment]
> What if the native implementations, but not the web implementations, add support for synchronous readback on the main thread?

This sounds particularly un-nice.

> "a couple frames" to "~1ms" is quite a big pivot!
> ...
> 1ms is, IMO, absolutely not heavy enough to justify forcing async.

Still agreeing hard.

**Comment by Kangz** [comment]
> "a couple frames" to "~1ms" is quite a big pivot! Even on a high-end phone with a 120hz display, that's still only an 8th of a frame.

If you keep flushing the GPU pipeline, yes you get ~1ms, but then the GPU spends the majority of its time idle because there is no pipelining.

**Comment by valadaptive** [comment]
The GPU also spends the majority of its time idle if you hit your frame budget. It seems like your objection here is not actually that synchronous readback causes any jank or inevitably results in a worse user experience, but that it feels icky for the main thread to ever be idle.

I get that it *feels* wrong to "block" the main thread, but the definition of "blocking" is [not as objective as it seems](https://blog.yoshuawuyts.com/what-is-blocking/). Is writing to a file descriptor a "blocking" action? It's IO-bound. What about writing to stdout?

The Web platform's reluctance to perform any "blocking" operations whatsoever on the main thread seems to have led to no small number of gripes over the years, and for good reason. In particular, the decision to [forbid `Atomics.wait` on the main thread](https://github.com/WebAssembly/threads/issues/177) has caused a lot of developers a lot of real pain.

> Of course the rest of the code often needs to be adapted to handle the asynchrony because it might try to do something in the meantime, and there's a development cost to that.

"A development cost" is underselling things a fair bit. As pointed out in the `Atomics.wait` GitHub thread, there is *no programming model* for developing applications where every function is re-entrant. [The Rust standard library itself](https://github.com/rust-lang/rust/blob/f45631b10f2199a7ef344a936af9eb60342c16ee/library/std/src/sys/wasm/alloc.rs#L71), as well as Emscripten, are forced to busy-wait because the Web platform folks have deemed it not "best practices" to yield from the main thread. If *Emscripten* can't figure it out, who can?

But OK, let's assume the worst[^1] and consider the impact of the main thread being blocked for several frames. Let's say some dev forgot to consider an edge case and allowed a certain operation to take way too long, or it's running on lower-end hardware than was envisioned.

Which would you rather have:
- The webpage freezes for several frames, worsening the UX.
- Everyone gets a window of several frames to accidentally stress-test whether your codebase is *really* reentrant.

[^1]: If you extend things to atomics, as referenced above, things become worse: there is the possibility of deadlocks. But browsers already have machinery in place if code spends too much time in a loop. JavaScript, being Turing-complete, can already block the main thread, and indeed I've accidentally written a few infinite loops myself. I've also seen an "a script is slowing down the page" prompt on many a production website.

**Comment by kainino0x** [comment]
> What if the native implementations, but not the web implementations, add support for synchronous readback on the main thread?

This is basically the current plan - we're adding a synchronous-wait primitive to webgpu.h. In Wasm, it will be supported when JSPI (or Asyncify or similar emulation) is available.

> 1ms is, IMO, absolutely _not_ heavy enough to justify forcing async.

The 1ms isn't the justification for async (it probably could be even faster than that - I think non-Chrome browsers do better than 1ms already). That's just the overhead - the actual justification is there's an arbitrary amount of work that's already been queued to the GPU that must complete before the mapSync/readPixels completes. It could be trivial, sure, but it could also take a long time (especially on a low end phone).

That said, while it's technically possible for the webpage (most importantly touch-scrolling) to remain responsive while a lot of WebGPU work is queued, it's often going to be the case that the webpage or even the entire browser stops displaying frames while it completes. In which case the mapSync is probably not making anything worse.

Anyway, TBH, I think we're rehashing all the same arguments. I still personally am on the side of adding mapSync in all contexts, partly because of historical precedent with WebGL and 2D canvas, and partly because of the thing in the last paragraph. But regardless of whether mapSync happens, JSPI is on its way, and we'll make the best we can with that.

**Comment by kainino0x** [comment]
I'll tentatively triage this into Milestone 1 as it's clearly still a hot topic. Can't guarantee it will see any real progress though.

**Comment by litherum** [comment]
Speaking purely as as an independent observer here:

If you don’t add mapSync on the main thread, the drawbacks are that some applications will use mega-janky toDataURL workarounds, while others just won’t be ported to WebGPU (or possibly even the entire web platform) at all.

If you do add mapSync to the web platform, the drawbacks are that some apps which _don’t_ need mapSync may elect to use it anyway, thereby resulting in unnecessarily-janky apps.

If I had to choose between those two outcomes, I’d personally choose to have those apps supported on WebGPU, and choose a scary name for mapSync to try to make it clear that authors should avoid using it if possible. Personally, I’d care way more about seeing Jet Set Radio or whatever running in a browser on WebGPU than worrying about how janky some one-off app someone wrote for who-knows-what reason is.

**Comment by juj** [comment]
Here are some notes that I can pick up from the discussion so far:

a) In the previous comments, Emscripten's `Asyncify` feature has been mentioned a couple of times. While it is a feat of engineering, Asyncify really never was a production quality feature, due to the dramatic collateral problems it has in increased code size, wasm build times, reduced runtime performance, and re-entrancy challenges. Unfortunately Asyncify results in trading problems with other problems.

b) JSPI is the more robust incarnation/evolution of Asyncify (obsoleting Emscripten's original Asyncify model), and fixes many of its problems. However, even then, problems remain to be resolved with JSPI:

- The one problem it does not fix is the re-entrancy challenge, which was mentioned already [above](https://github.com/gpuweb/gpuweb/issues/2217#issuecomment-1961983650), picturing quite well how this creates a surface area for novel bugs as opposed to a simple-to-reason known quantity that mapSync() would be. The problem here is that when the call stack is suspended to wait for the async operation to finish, the general browser/JS event loop will continue running. This means that the developer will need to have global awareness of the whole codebase/site in question to know which JS events they can allow to keep resolving, and which they should hold in a queue to resolve only later after the paused async operation has completed (and whether there are transitive ordering guarantees between such events that require holding other events). This makes composability of software libraries more difficult, as a middleware rendering library may have a hard time coordinating such global information with the embedder, or other JS libraries that fire JS events.

- Another problem here is that JSPI interacts with the presentation model of the DOM Canvas element. Canvas element is specced to present implicitly after finishing any event handler that performed draw operations to the front buffer. In JSPI's case there is currently no special handling to this: the event handler finishes at the moment when the JSPI pause occurs, not when the paused operation resolves. This means that sites that would utilize JSPI'd await mapAsync() would need to render into an offscreen temp render target (to-be-blitted-on-screen at the very end) so that they would not risk presenting intermediate rendered data. This could lead to requiring an extra fullscreen blit in codebases (unless a renderer can sink it into some other backbuffer draw _and_ statically reason that their last blit will never straddle mapAsyncs in the middle of that final presentation blit at a bad time)

- Third question is how the JSPI'd pause-resumed event handlers will interact with requestAnimationFrame() timings. It is common for renderers to need to perform several sync buffer maps within a single frame, not necessarily just one. There could be dozens or more of such mappings per frame, especially at scene load times or at large dynamic object instantiation times. With JSPI, This would be akin to having a rAF() event handler first perform some WebGPU ops, and then several subsequent "setTimeout() style" events perform pieces of more rendering, before the next rAF() to come.
\
The question to ponder here is if this kind of operation will be optimally performant with the rAF()/canvas vsync timing models? For example, if the timings don't quite match up, the next rAF() might arrive while in the middle of a stream of these JSPI paused mapAsync() calls: the codebase would likely need to detect that and skip out from handling anything in that rAF() (i.e. it is detecting re-entrancy). But how will that affect frame presentation timing? Is there danger of performance being unduly pessimized? (e.g. to renderers that are close to their frame time budget already?) Some guidance will likely be needed in this area to confirm this won't be a possible can of worms.

c) There was some mentions above about having waitable map()ing support rather than synchronous mapping support. In a hypothetical waitable map() scheme, user could initiate an asynchronous map(), but then at some point later (while still computing in that same event handler), they could decide to either poll if completion has occurred, or to synchronously pause to wait for completion. This kind of waitable model could be useful, since it would allow renderers to initiate several map requests at once, and then wait for those to resolve in first(), any() or all() fashion. That would allow renderers opportunities to sink waits and perform other computations meanwhile. (maybe some renderers might be able to issue maps at beginning of frame, and then resolve those at the late stages of frame, or similar?)

If there would be a sequential completion guarantee for multiple mappings, then waiting for any() would be trivially easy (just wait in order of submission, e.g. wait for first()). Though if the completion order might be arbitrary, then this would result in the select problem. (this could be solvable via a SAB futex wait style method)

d) As a followup to the above: at this point conversation has revolved around the async->good, sync->bad dogma. It is easy to slip from this to think async->fast, sync->slow by construction. But there is no precedent to show that this would actually be the case. It is very possible that such a waitable mapping function, or an outright mapSync() function could actually result in a higher throughput/shorter CPU time spent in a renderer. Just because the maps resolve asynchronously via the JS event queue does not mean that these mappings would be fast - on the contrary, they could be pretty slow in comparison to a synchronous/waitable map due to JS event queue processing latency and computing context management overhead. But since there doesn't exist support for sync/waitable maps in browsers, there are no current benchmarks to verify this behavior.

To this effect, I would recommend browsers would add mapSync() support behind a flag e.g. in Nightly/Canary builds, so that WebGPU renderer experimentors would be able to measure the cost of the async mapping model to enable validating that there is no overhead to performing asynchronous mapping. It is known that async mapping requires utilizing more GPU memory for round robin dynamic buffers compared to being able to fence+map when unused. How much more? Well, that is harder to analyze. A behind-a-flag mapSync() would allow measuring this overhead in renderers as well. (in codebases that are able to do both async and sync maps)

I would say that for example, if waitable/sync mapping would give a +x% more FPS or -y% less GPU memory used by a renderer compared to async mapping, that would give a new angle to look at the dogmatic/pragmatic axis of sync vs async computation.

e) Strong +1 to adding mapSync() support permanently to Workers.

f) Has there been consideration to add a `getBufferSubData(dsttypedarray, offset, size)` style of readback API (i.e. synchronously read bytes from a buffer without persisting a mapping to it)?

This kind of API might be faster than a JSPI'd mapAsync()+read+unmap, e.g. by virtue of avoiding JSPI, and it could even be faster than a mapSync()+read+unmap by avoiding an extra JS->Wasm copy.

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2024-03-13#timebox-mapsync-on-workers---and-possibly-on-the-main-thread-2217">GPU Web CG 2024-03-13 Atlantic-time</a></summary>

* KN: I’d like to take the temperature on this proposal. Start with taking the temperature on mapSync on workers.
    * People keep asking for it, and as far as I know we've never formally asked the TAG for input on it (only relayed opinions from folks at the browsers). If there is general interest from the CG, I would like us to ask the TAG for review on a proposal defending it (many points have been raised in favor of it, and I honestly think synchronous GPU readbacks are a pretty trivial hazard).
    * Only negative feedback we got on this was someone from Google not wanting worker-specific APIs.
    * Think this would make a lot of people happy.
        * (BTW, there's [been some noise](https://github.com/WebAssembly/threads/issues/177) pushing for Atomics.wait on the main thread too, which I'd love to see, but it's much bigger.)
* KG: +1 to supporting this on workers.
* MW: catching up on the convo, but why do we need this, as opposed to worker waiting asynchronously?
* KN: couple reasons. Pure JS - can write logic in a way that doesn't require reentrancy. Doesn't force you to yield. Can start using it in requestAnimationFrame. Don't know whether we necessarily want to encourage this. It makes it possible to do mapping from Wasm without yielding - proposal getting standardized, JSPI (JS Promise Integration), still has a bunch of reentrancy problems. Very difficult in C++ code - these problems are major. On workers, there's no harm in allowing this.
* MW: initial input - seems fine with workers, not sure about the main thread.
* KN: great. Do we think we need to ask the TAG about mapSync on workers?
* KR: think we should.
* CW: think we should ask TAG. If they take too long to answer, can push forward.
* KG: imagine that if standards folks say OK, that's a strong enough signal.
* BJ: WebGPU wouldn't be the first API to add blocking support on workers. I have a comment on here from 2 years ago about FileSystemAccessHandle and sync operations. Think talking with 
* KR: Would this enable WebGPU in AudioWorklet?
    * BJ/KN: Would go a long way but other problems
* KR: Many customers that need the data back on CPU “now”.
* GT: since I write a bunch of tools to patch the API, seems impossible in workers. Wouldn't be able to block WebGPU, disable features - extension privacy issue. Can do this in main thread. If we add a feature to workers and everyone switches over to workers, tools will stop working. Talk with extension folks?
* KG: I'm with you. We should talk with folks who spec web-extensions. Not surprised, also not excited about this difficulty.
* KN: have work to do to bring this to TAG.
* CW: ok, so discussions with (1) doing this in workers (2) what happens if you do it in rAF, and block (3) etc. Think we should start a proposal doc.
* KG: welcome to put this in proposals. I'm a bit surprised you're concerned about this since we have readPixels, getImageData on 2D canvas, etc.
* BJ: think would be useful to present as much real-world data as possible. Hope group wouldn't object to browsers wanting to implement this behind a flag so we can gather some real data.
* CW / KG: do whatever you want behind a flag.

</details>

**Comment by juj** [comment]
> Start with taking the temperature on mapSync on workers.
> People keep asking for it, and as far as I know we've never formally asked the TAG for input on it
> Only negative feedback we got on this was someone from Google not wanting worker-specific APIs.
> KN: great. Do we think we need to ask the TAG about mapSync on workers?

There has been a historical precedent towards synchronous APIs being allowed in Workers. (So I expect TAG would be ok with this, unless there has been a change in direction somehow)

Examples:
 - [FileReaderSync](https://developer.mozilla.org/en-US/docs/Web/API/FileReaderSync) is available in Workers, which allows synchronously reading a `Blob` of a `File` into an `ArrayBuffer`,
 - the synchronous [Atomics.wait()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/wait) operation that was already mentioned,
 - synchronous XHR in Workers (although there is some pressure to deprecate the XHR API in whole, and synchronous Fetch was never standardized)

--------------------

Btw the discussion of synchronous work in Workers reminded me of **a priority inversion problem**. Maybe a bit tangential, but maybe worth mentioning here: An important detail about browser APIs in Workers \*plus\* SharedArrayBuffer is the fact that SAB makes it possible to observe in user JS code if such operations ***require assistance/forward progress*** from the main browser thread in order to advance. Such work requests can result in priority inversion deadlocks. Some examples:
 - [Firefox used to have a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1049091) that if any Worker called `console.log()`, then that Worker would internally synchronously pause to wait for the main browser thread to acknowledge that it has processed the logging. If the SAB code on the main JS thread was simultaneously spinwaiting to receive a result from the Worker, the main thread and the Worker would deadlock together.
 - all browsers today have a "problem" that Workers cannot spawn Workers without forward progress help from the main thread. Hence, if a Worker was spawning a Worker while the main thread was spinwaiting on SAB for something from the Worker, these would deadlock.
 - likewise, main browser thread cannot spawn Workers without help from the main thread itself, resulting in [a rather nasty can of worms](https://github.com/emscripten-core/emscripten/blob/29be801344a9bf726abc2e146dbd3dfed4a86c9b/src/settings.js#L1615-L1635) for SAB users.
 - [a test case about sync XHRs in Workers](https://github.com/azriel91/worker_sync_xhr) showcases how the XHR API (in some unspecified browser) has this property. Worker launches a sync XHR, which requires help from the main browser thread to progress. If the Worker was holding a user lock on a SAB and the main thread happened to spinwait on that lock, then deadlock would occur.

(one might argue that these are examples of browser vendors writing the very type of poor practices code into browser codebases, that they are trying to keep bad JS developers from writing, but best not go there...)

It is good to note that before the advent of SharedArrayBuffer, none of this was a problem, since there was no synchronous shared state between the Worker and main thread. So even if any APIs in Workers might synchronously halt to wait for a lock/event from the main thread, it would never be possible to functionally observe.

There was little traction in attempting to update wording on already shipped W3C specs about this, but rather these types of issues have been handled as individual bugs against browsers. For example [we used to proxy all console.logs in Emscripten manually](https://github.com/emscripten-core/emscripten/commit/c32101013cf1b54f294670cc41bed6720b44da2f) over to the main thread to log, until later when Firefox no longer had the problem.

The point of this is that when specifying behavior of WebGPU in Workers, it would be good to have a paragraph to mention that implementations would be required to be able to make forward progress without synchronous help from the main browser thread. Otherwise there is a possibility of priority inversions to occur when talking about Workers, and SAB is in town.

Of course this is not something that would be a potential challenge only with synchronous operations like `mapSync()`, all other operations in the spec may also have this hurdle - though I was just reminded of that possibility in this context.

**Comment by Morglod** [comment]
For everyone trying to do same thing with native webgpu. macos m1. Tried many variants with wgpu-rs, found that mapAsync is terrible slow and its just wgpu-rs implementation issue. Only one good variant is to take google's dawn implementation, make ring buffers for reading and do this (every frame):

<details>
<summary>cpp code with glfw and dawn</summary>

```cpp
cpu_buffer = buffers_for_reading.next_buffer();

encoder.copyBufferToBuffer(gpu_buffer, cpu_buffer);
queue.submit( ...encoder.finish()... );

// map and wait
bool done = false;
wgpuBufferMapAsync(cpu_buffer, mode, offset, size, [](WGPUBufferMapAsyncStatus status, void* user_data) {
    (*((bool*)user_data)) = true;
}, &done);

while(!done) {
    glfwPollEvents();
    wgpuInstanceProcessEvents(instance);
}

// getConstMappedRange and do smth with data

cpu_buffer.unmap();
```

</details>

Also for everyone fighting with wgpu-rs -> dawn migration. Google decided to have "undefined" flags not as 0 and have more validations on it. So You may get strange errors and you should solve it by specifying all descriptor fields.
I was stuck with this: `WGPURenderPassColorAttachment { .depthSlice = WGPU_DEPTH_SLICE_UNDEFINED }`

**Comment by toji** [comment]
I think it's worth sharing with the working group that the Chrome team has had a confirmation that at least one well known, real world application has independently invented the hack that I detailed in https://github.com/gpuweb/gpuweb/issues/2217#issuecomment-1067016745 and is using it in the wild. I'm not going the name the app here without their permission, but felt like this was noteworthy enough that it needed to be communicated here.

The application was ported from a WebGL backend, and according to their engineer "In spite of its awfulness, I am grateful that the hack exists, because we would not have been able to ship WebGPU support without it!" They've also indicated that while they are willing to refactor their code it would be a significant effort. Specifically: "The lack of a drop-in-replacement for synchronous readPixels would have meant that we simply couldn't have used WebGPU for rendering, at least not without quite a lot (a year?) more work, if not for the canvas hack."

While I understand and broadly agree with the intent behind avoiding synchronous operations on the main thread, I also am deeply saddened by the fact that we have developers who are knowingly choosing to implement a terrible hack that they acknowledge performs poorly because it's the most practical way to ship with WebGPU, something that they were very eager to do!

Surely we can provide a better, cleaner, and far more efficient path than this.

**Comment by aringleinfigma** [comment]
Hi! The app mentioned above is Figma and I'm an engineer working on our WebGPU-based rendering backend. The comment covers our situation well, so just confirming that a mapSync API would be extremely helpful!

**Comment by Morglod** [comment]
Message to webgpu working group: please just implement readSync. It was actually the fastest way to read data on native OpenGL each frame (i benchmarked it a lot) (same as BufferDataUpdate). Inside implementation you could have this ring buffers and other things that you force to use on user side. All this async mappings is bad because no one knows how each vendor implement their backend. Vendors should know how to sync their stuff efficiently.

**Comment by ambroser53** [comment]
Is there any idea of intention for work on this? 

I'm having the exact same issue as you @Morglod where I'm seeing huuuuge slowdown when using mapAsync on Mac (and only on Mac) for downloading from gpu to cpu. Do you have a work around that reliably works? I can't really implement async at all within my usecase because it's deep within an entirely synchronous low-level dll. I'm using dawn not wgpu-rs

**Comment by kainino0x** [comment]
mapSync is not meant to be a solution to excessive latency, but to other problems relating to needing to yield to the JS event loop. In principle, if mapAsync is slow, mapSync will be slow too. (There would be slight scheduler overhead to an async task, but this should be much smaller than the excessive latency we have in Chromium right now which I believe is happening at a different layer. We are actively investigating that, but we don't have an answer yet.)

**Comment by ambroser53** [comment]
Thanks @kainino0x, I've come to the same realisation at the same time. Are you referring to https://dawn-review.googlesource.com/c/dawn/+/254435 by any chance?

**Comment by kainino0x** [comment]
@ambroser53 Not specifically. I'm not familiar with that change but it sounds like it affects the memory access speed after mapping on Vulkan, but we know the initial time to map is also long, on all platforms.

**Comment by Morglod** [comment]
@kainino0x Because optimized sync read/write methods was always faster than mapping in any API. And will be I suppose always faster. Probably some async read/write will be better, but not mapping. _Cant understand why its so hard to add ancient working technology_

**Comment by kainino0x** [comment]
The `mapSync` proposal here is exactly the same as `mapAsync` except that it blocks until the buffer is mapped rather than returning a promise. I think you are talking about some kind of simultaneous / unsynchronized / manually synchronized sharing between the CPU and GPU? That is not under consideration because of the Web's strong security requirements.

**Comment by Morglod** [comment]
@kainino0x The only purpose of mapping is to read or write data stored inside buffer. There is **!! no need !!** in mapping, when you can do smth like `readSync` or `writeSync`. Which was and will be always faster. **!! It is supported now in WebGL !!**, there is nothing about security. `gl.bufferSubData` and `gl.getBufferSubData`.

If the problem is that you want to have verbose API with explicit synchronization, than this writes/reads could be added to command queue.

**Comment by kainino0x** [comment]
@Morglod WebGPU has `GPUQueue.writeBuffer` and `GPUQueue.writeTexture` already. There isn't an equivalent of `gl.getBufferSubData` though.

**Comment by mwyrzykowski** [comment]
Indeed we already have synchronous writes via above mentioned `writeTexture` and `writeBuffer`.

Can we better understand the use case for synchronous buffer reading? That is always going to be slow as it blocks the GPU pipeline and removing an async dispatch should be measured but is likely a micro perf gain as @kainino0x notes in https://github.com/gpuweb/gpuweb/issues/2217#issuecomment-3130695173

**Comment by valadaptive** [comment]
> Indeed we already have synchronous writes via above mentioned `writeTexture` and `writeBuffer`.
> 
> Can we better understand the use case for synchronous buffer reading? That is always going to be slow as it blocks the GPU pipeline and removing an async dispatch should be measured but is likely a micro perf gain as [@ kainino0x](https://github.com/kainino0x) notes in [#2217 (comment)](https://github.com/gpuweb/gpuweb/issues/2217#issuecomment-3130695173)

Beyond the speed issue, a ton of people have dropped into the thread detailing their use cases (including a Figma engineer who had to resort to an ugly hack to ship WebGPU!)

Inserting event-loop yield points into your code isn't just something you can do without thinking. Everything that needs to become async has to think about reentrancy:

> > Of course the rest of the code often needs to be adapted to handle the asynchrony because it might try to do something in the meantime, and there's a development cost to that.
> 
> "A development cost" is underselling things a fair bit. As pointed out in the `Atomics.wait` GitHub thread, there is _no programming model_ for developing applications where every function is re-entrant. [The Rust standard library itself](https://github.com/rust-lang/rust/blob/f45631b10f2199a7ef344a936af9eb60342c16ee/library/std/src/sys/wasm/alloc.rs#L71), as well as Emscripten, are forced to busy-wait because the Web platform folks have deemed it not "best practices" to yield from the main thread. If _Emscripten_ can't figure it out, who can?

It's very easy to introduce hard-to-trigger and hard-to-test race conditions with async. And the guts of your fancy rendering or GPGPU engine is the worst possible place to force an event loop yield.

@juj has [some good commentary](https://github.com/gpuweb/gpuweb/issues/2217#issuecomment-1995237435) on this point.

**Comment by valadaptive** [comment]
I will say that it's quite frustrating to have several people, over the course of years, drop into this thread detailing their use cases, and still have spec maintainers and implementers asking "can you explain your use case?" I'm sure there's a lot of discussion going on internally, but to outside observers who have to figure out how to actually use these APIs, it's quite opaque.

**Comment by Morglod** [comment]
@mwyrzykowski Well, same as for reading through mapping, but without overhead of mapping. Its per-frame gpu computations.

When you want to have **SYNC READ RIGHT NOW**, with current api, you should do a lot of unnecessary stuff with unnecessary overhead.

_To be honest it feels like now you are working on inventing new excuses rather than on solving this issue._

**Comment by mwyrzykowski** [comment]
> [@mwyrzykowski](https://github.com/mwyrzykowski) Well, same as for reading through mapping, but without overhead of mapping. Its per-frame gpu computations.

@kainino0x mentioned it in https://github.com/gpuweb/gpuweb/issues/2217#issuecomment-3137508305 but in this context the words 'read' and 'mapping' mean the same.

To help illustrate the difference, the current API is implemented something like (very high level pseudo-code):
```
mapAsync(buffer) {
  gpuDriver.callbackWhenAllWorkOnResourceFinishes(buffer, () => {
     fulfill JS promise with contents of 'buffer'
   });
}
```
it sounds like there is a request for:
```
mapSync(buffer) {
  semaphore s;
  gpuDriver.callbackWhenAllWorkOnResourceFinishes(buffer, () => {
     synchronously return from JS function with contents of 'buffer', which is possible due to waiting on a semaphore
     s.signal();
   });
  s.wait();
}
```
as previously pointed out this will only potentially mitigate a thread switch.

If I'm understanding correctly, it sounds like making sync JS / wasm to become asynchronous is more than a non-trivial effort?

I think we all legitimately want to understand how to help solve this, given the demand :)


Is there need for this outside of workers? For workers, this seems completely reasonable and quite simple for UAs to implement as well. 

If this is really desired on the main thread, we would need to do some more thinking due to blocking of the main thread.

**Comment by Kangz** [comment]
I'd love to not need to lock this thread. Please stay civil and assume good faith effort from all parties involved. The request for more info is to 1) figure out if we can find another alternative (sync waiting is a no-no in guidelines for new Web APIs) 2) which workers (or main thread) need this.

Also don't underestimate the effort that this requires to spec and implement. At least in Chromium making JS able to synchronously wait on stuff for WebGPU in GPU process side is a complex change, my guess is 1-2 months of engineering. That's even before any implementation of a `mapSync`. There's also a matter of prioritization especially at times where Firefox and WebKit a reaching a first shippable version of WebGPU.

**Comment by Dinnerbone** [comment]
> sync waiting is a no-no in guidelines for new Web APIs

I think this, fundamentally, is the issue that my project has with this. _Existing_ web APIs, notably WebGL, allow for this, and so existing applications require this. If we want to start switching to WebGPU for more capabilities, performance or whatever - we need to at least have the same capabilities that WebGL allowed for, which means a sync readback of buffers.

For context, our application is an emulator for Flash (.swf) files, which can be wholly emulated right now using webgl. Flash allowed, in fact encouraged, sync readback and therefor we need to allow for it too - which means switching to webgpu is off the table for us, despite in theory being an upgrade in every other way. Our use case then isn't strictly about mapping a buffer sync, but doing a single readback sync.

It's a very good thing to not block the main thread or stall the gpu in new applications, I completely agree - but sometimes it _is_ needed, and we should not dismiss decades of history that used it because by todays standards they were doing something suboptimal.

**Comment by fanciful-marmot** [comment]
I've spent some time trying out the recent [experimental mapSync in workers](https://developer.chrome.com/blog/new-in-webgpu-145#experimental_synchronous_buffer_mapping_in_workers) support in Chrome and wanted to give some feedback on it.

MapSync (even only in a worker) would have been _very_ useful to us if it had existed. In our case, we have some C++ code that we want to adapt to compile with emscripten and render with WebGPU. Since this code originally ran on desktop, it has the assumption that synchronous readback from the GPU is possible baked into it. This assumption can be a lot of work to refactor out. In our case it was mostly around GPU-based picking. With no way of doing synchronous mapping, our only option was to invest significant effort in a refactor, or fallback to CPU picking. CPU picking is also synchronous, blocking, and can be slow. Having the possibility of using mapSync in a worker context would have saved us from this and made it easier to move forward.

One note on the API: In our case, there are multiple buffers that need to be mapped back before we can proceed. If using `mapAsync`, it's possible to make all the `mapAsync` requests at once and then do something with all the results when available using `Promise.all()`. Using `mapSync`, you can't make the next mapping request until the first one resolves.

Roughly,
```javascript
// Faster, parallel requests
await Promise.all([buffer1.mapAsync(), buffer2.mapAsync(), buffer3.mapAsync(), buffer4.mapAsync()])

// Slower, sequential requests
await buffer1.mapAsync()
await buffer2.mapAsync()
await buffer3.mapAsync()
await buffer4.mapAsync()

// About the same as sequential. No option for issuing all requests upfront and then a blocking wait
buffer1.mapSync()
buffer2.mapSync()
buffer3.mapSync()
buffer4.mapSync()
```

I'm unsure how it all works internally, but if it's faster to know upfront about the mapping requests before waiting on the GPU, then being able batch issue the mapping requests would be helpful for our use case. There may be other solutions, like writing all the data into 1 big buffer instead of 4 smaller ones, but I wanted to put this out there.

```javascript
// A questionable (but helpful) API.  Equivalent to the await Promise.all([...]) above, but sync
const allResults = GPUBuffer.batchMapSync([buffer1, buffer2, buffer3, buffer4])
```

Even as the API is, this still would have been very, very helpful to have had.

**Comment by greggman** [comment]
Did you consider refactoring to use slightly old picking data?

## original picking code

* render to `pickingData`
* synchronously map `pickingData`
* synchronously use `pickingData`

## new picking code

* have `pickingData` and `pendingPickingData`
* render to `pendingPickingData`
* asynchronously map `pendingPickingData` `.then() { 
    [pendingPickingData, pickingData] = [pickingData, pendingPickingData]; // swap refs
  });`
* synchronously use `pickingData`

This leaves your main loop synchronous and all that happens is the data is a moment off. Bonus, you don't stall the rendering pipeline with a sync map so your app feels more responsive. It's also often a very minor refactor.

**Comment by fanciful-marmot** [comment]
Yes, that is basically the strategy we ended up with for most picking since we often want to cache the pick data for the frame for faster lookups if the view hasn't changed. We would do this even if it was possible to sync map but there is a complication if you can only do async mapping. There are also some picking paths like deep picking which hits deeper than the surface where we can't cache everything.

For caching the pickingData, there is a window of time where you know that pickingData is out of date and that a request is in flight to get the latest pickingData. If a user action that requires picking data (like a click) happens in that window then what do you do? You know that the data you have is wrong so if you base some action on that then it might be wrong. You have a few options:
* Process the event with out of date pickData. You might hit the wrong thing which isn't good depending on what you're trying to do.
* Just return no-hits since you can't know if the hit will be accurate. Might be fine for a mousemove where another event is likely to come in but not great for something like a click
* You can delay processing the action until the latest data is ready. That means you have to queue up input events (or anything that might need a hit test) and delay processing them if you ever have out of date pick data, which is a larger change.
* You can fallback to CPU hit testing. This also blocks and is slower than mapSync. It's also more complicated since you now have 2 different picking mechanisms to keep identical.

For the third option, this is basically the same situation you end up in with sync mapping except with sync mapping a user action will only wait for the data if it's needed (when the mapping is triggered). If you can only do async mapping then you have to know before processing the user action whether or not it will need the pickingData, so that you can decide whether to delay it or not. Otherwise you have to delay all actions until pickData is ready

In practice this window of time may be quite small, but you can get into situations where one action will invalidate the pick data with a camera move, the next action will definitely need picking data, and these will happen very fast. If these bugs do happen, they're a nightmare to track down. Having the synchronous path available can give much more confidence that the ported code is working properly and then we can work towards the async path from there
